crypto/asn1/d2i_pr.c

```diff
@@ -78,13 +78,40 @@ EVP_PKEY *d2i_PrivateKey(int type, EVP_PKEY **a, const unsigned char **pp,
  * type
  */
 
+static EVP_PKEY *key_as_pkcs8(const unsigned char **pp, long length)
+{
+    const unsigned char *p = *pp;
+    PKCS8_PRIV_KEY_INFO *p8 = d2i_PKCS8_PRIV_KEY_INFO(NULL, &p, length);
+    EVP_PKEY *ret;
+
+    if (p8 == NULL)
+        return NULL;
+
+    ret = EVP_PKCS82PKEY(p8);
+    PKCS8_PRIV_KEY_INFO_free(p8);
+
+    if (ret != NULL)
+        *pp = p;
+
+    return ret;
+}
+
 EVP_PKEY *d2i_AutoPrivateKey(EVP_PKEY **a, const unsigned char **pp,
                              long length)
 {
     STACK_OF(ASN1_TYPE) *inkey;
     const unsigned char *p;
     int keytype;
+    EVP_PKEY *ret = NULL;
+
+    ret = key_as_pkcs8(pp, length);
```

Comment:
While I understand the cleanliness of having it a separate function, it also makes you lose an important distinction: did it fail because it's not PKCS#8 (i.e. `d2i_PKCS8_PRIV_KEY_INFO()` fails) or because the contents of the PKCS#8 blob are unsupported (i.e. `EVP_PKCS82PKEY()` fails)?  I think that should be a deciding factor to try something else or not.
To achieve that, making PKCS#8 parsing a separate functions doesn't seem like the best choice...

Reply:
Fair point. Thanks!